DatadogTrace/Sources/OpenTelemetry/OTelSpan.swift

```diff
@@ -0,0 +1,162 @@
+    var kind: OpenTelemetryApi.SpanKind
+    let nestedSpan: DDSpan
+    let tracer: DatadogTracer
+    let queue: DispatchQueue
+
+    /// `isRecording` indicates whether the span is recording or not
+    /// and events can be added to it.
+    var isRecording: Bool
+
+    var status: OpenTelemetryApi.Status {
+        get {
+            fatalError("Not implemented yet")
+        }
+        set {
+            fatalError("Not implemented yet")
+        }
+    }
+
+    /// `name` of the span is akin to operation name in Datadog
+    var name: String {
+        get {
+            queue.sync {
+                _name
+            }
+        }
+        set {
+            queue.sync {
+                guard isRecording else {
+                    return
+                }
+                _name = newValue
+            }
+            nestedSpan.setOperationName(name)
+        }
+    }
+
+    init(
+        attributes: [String: OpenTelemetryApi.AttributeValue],
+        kind: OpenTelemetryApi.SpanKind,
+        name: String,
+        parentSpanID: OpenTelemetryApi.SpanId?,
+        spanContext: OpenTelemetryApi.SpanContext,
+        spanKind: OpenTelemetryApi.SpanKind,
+        startTime: Date,
+        tracer: DatadogTracer
+    ) {
+        self._name = name
+        self._status = .unset
+        self.attributes = attributes
+        self.context = spanContext
+        self.kind = kind
+        self.isRecording = true
+        self.queue = tracer.queue
+        self.tracer = tracer
+        self.nestedSpan = .init(
+            tracer: tracer,
+            context: .init(
+                traceID: context.traceId.toDatadog(),
+                spanID: context.spanId.toDatadog(),
+                parentSpanID: parentSpanID?.toDatadog(),
+                baggageItems: .init()
+            ),
+            operationName: name,
+            startTime: startTime,
+            tags: [:]
+        )
+    }
+
+    // swiftlint:disable unavailable_function
+    func addEvent(name: String) {
+        fatalError("Not implemented yet")
+    }
+
+    func addEvent(name: String, timestamp: Date) {
+        fatalError("Not implemented yet")
+    }
+
+    func addEvent(name: String, attributes: [String: OpenTelemetryApi.AttributeValue]) {
+        fatalError("Not implemented yet")
+    }
+
+    func addEvent(name: String, attributes: [String: OpenTelemetryApi.AttributeValue], timestamp: Date) {
+        fatalError("Not implemented yet")
+    }
+    // swiftlint:enable unavailable_function
+
+    func end() {
+        end(time: Date())
+    }
+
+    func end(time: Date) {
+        queue.sync {
+            guard isRecording else {
+                return
+            }
+            isRecording = false
+
+            // Attributes maps to tags in Datadog
+            for (key, value) in attributes {
+                switch value {
+                case .string(let value):
+                    nestedSpan.setTag(key: key, value: value)
```

Comment:
question/ Why this `queue`? The `nestedSpan.setTag(key:value:)` is already thread safe.

Reply:
Yes, there is no need and perhaps it can lead to a deadlock if same queue is used.

Refactored.